fms/models/llava_next.py

```diff
@@ -0,0 +1,586 @@
+    def from_config(cls, config: LlavaNextConfig) -> "LlavaNext":
+        return cls(config)
+
+    def get_config(self) -> LlavaNextConfig:
+        return self.config
+
+    def reset_parameters(self):
+        nn.init.xavier_uniform_(self.image_newline.data)
+        self.langauage_model.reset_parameters()
+        self.vision_tower.reset_parameters()
+        self.multi_modal_projector.reset_parameters()
+
+    def post_init(self):
+        self.language_model.post_init()
+        self.vision_tower.post_init()
+
+    def unpad_image(self, tensor, original_size):
+        if not isinstance(original_size, (list, tuple)):
+            original_size = original_size.tolist()
+        original_height, original_width = original_size
+        current_height, current_width = tensor.shape[1:]
+
+        original_aspect_ratio = original_width / original_height
+        current_aspect_ratio = current_width / current_height
+
+        if original_aspect_ratio > current_aspect_ratio:
+            scale_factor = current_width / original_width
+            new_height = int(round(original_height * scale_factor, 7))
+            padding = (current_height - new_height) // 2
+            unpadded_tensor = tensor[:, padding : current_height - padding, :]
+        else:
+            scale_factor = current_height / original_height
+            new_width = int(round(original_width * scale_factor, 7))
+            padding = (current_width - new_width) // 2
+            unpadded_tensor = tensor[:, :, padding : current_width - padding]
+
+        return unpadded_tensor
+
+    # TODO: fix graph break in the HF impl here
+    def select_best_resolution(self, original_size, possible_resolutions: list):
+        if not isinstance(original_size, (list, tuple)):
+            original_size = original_size.tolist()
+
+        original_height, original_width = original_size
+        best_fit = None
+        max_effective_resolution = 0
+        min_wasted_resolution = float("inf")
+
+        for height, width in possible_resolutions:
+            scale = min(width / original_width, height / original_height)
+            downscaled_width, downscaled_height = (
+                int(original_width * scale),
+                int(original_height * scale),
+            )
+            effective_resolution = min(
+                downscaled_width * downscaled_height, original_width * original_height
+            )
+            wasted_resolution = (width * height) - effective_resolution
+
+            if effective_resolution > max_effective_resolution or (
+                effective_resolution == max_effective_resolution
+                and wasted_resolution < min_wasted_resolution
+            ):
+                max_effective_resolution = effective_resolution
+                min_wasted_resolution = wasted_resolution
+                best_fit = (height, width)
+
+        return best_fit
+
+    def image_size_to_num_patches(self, image_size, grid_pinpoints, patch_size: int):
```

Comment:
this can be rewritten without a single loop

Reply:
can you rewrite this using modular arithmetic?

Reply:
Done!